### ISSUE TITLE
[RFC] Classes and type definitions

### ISSUE THREAD

**Issue by faultyserver** [myst-lang/myst#16]
Right now, there is no way in the language to create a new data type, nor is there a way to add methods to native types (e.g., `List`).

In the previous iteration of the interpreter, the native types were treated as Modules in the Kernel that could be re-opened to add methods. This was _okay_, but meant that there wasn't a distinction between a module of methods and a module for a type. In fact, much like in Python, it was possible to do things like `List.each([1,2,3]) do ... end` as a replacement for `[1,2,3].each do ... end`. I didn't particularly like this, but it was a side-effect of how the semantics of function lookup and receivers were implemented.

In this new iteration, before simply copying over the old behavior, I'd like to rethink how type classes and type definitions can be implemented.

The default would be to essentially copy Ruby's implementation, where all types are `Class`es and new types are created with the `class` keyword:

```ruby
class Array
  # This would re-open the native Array class.
end

class NewType
  # This would define a new type call NewType.
end

# By default, classes have a `.new` method that creates a new instance of the type.
list  = Array.new
thing = NewType.new
```

In Ruby, classes are actually just modules with some added default behavior: the default implementations of `self.new`, etc., and the ability to actually allocate memory. This, to me, is one of the more difficult parts of Ruby for new programmers (especially those coming from static languages) to understand: that classes are mutable objects with no real special meaning (but actually with a little special meaning, then with metaclasses and all that other fun stuff).

The class implementation makes sense in Ruby as part of the rest of the language's design and values. However, I don't think it fits particularly well with what I see as the design and values of Myst : having minimal implicit behavior while retaining terseness and clarity.

I'd like this thread to suggest new and different ways of defining types (either from a syntactic or semantic angle) that can better meet that design goal. At this point, there is no bad suggestion; any and all input is appreciated. I have a few ideas that I'll throw out here soon. Voting with 👍  and 👎 is also appreciated :)

**Comment by faultyserver** [comment]
I hope to elaborate on this idea later. For now, I'm just posting the code to keep from forgetting it.

```ruby
type Person
  @name      : String
  @birthday  : DateTime
  @age       : Integer

  def greeting
    "Hello, " + @name
  end
end


jon = Person.new{ "Jon", D1995-11-14, 21 }
jon.greeting #=> "Hello, Jon"
```

**Comment by faultyserver** [comment]
```ruby
deftype Person
  name      : String
  birthday  : DateTime
  age       : Integer
end

module Person
  def greeting
    "Hello, " + @name
  end
end

static module Person
  def create_twin(p : Person, name : String)
    Person{ name, p.birthday, p.age }
  end
end


jon = Person{ "Jon", D1995-11-14, 21 }
twin = Person.create_twin(jon, "Jason")

jon.greeting #=> "Hello, Jon"
twin.greeting #=> "Hello, Jason"
```

I like this because it seems easy to implement :)

Another big advantage is a complete separation of data and functionality, while still allowing the nice `instance.method` and `Type.method` syntaxes. Perhaps some variation on `module Person` and `static module Person` to help reduce spelling errors would be good.

The `type` syntax is essentially `struct` from languages like C, VB, and Rust. They are simply a set of properties that define the components of the type. `module` and `static module` then define instance and "class" methods respectively. Again the added separation is advantage in my mind (no arguments over `def self.create_twin` vs. `class << self; def create_twin`, etc.).

The syntax for instantiation could use work. I don't mind the brace-initialization from C/C++ and Rust, but it feels out of place, especially with the braces already being used for both Map literals and block definitions. However, the brace-initialization does make pattern matching with custom types a little more feasible (see #11):

```ruby
def same_name?(Person{name: name}, Person{name: <name>})
  true
end

def same_name?(_, _); false; end
```

This seems closer to Elixir's style, which seems fitting, since Elixir also inspired Myst's pattern matching in general.

**Comment by faultyserver** [comment]
A potential iteration on the above:

```ruby
deftype Person
  name      : String
  birthday  : DateTime
  age       : Integer

  static def create_twin(p : Person, name : String)
    Person{ name, p.birthday, p.age }
  end
  # or
  defstatic create_twin(p : Person, name : String)
    Person{ name, p.birthday, p.age }
  end
  
  def greeting
    "Hello, " + @name
  end  
end
```

Here, `deftype`, `module`, and `static module` have all been compressed again (back to how classes are normally defined in Ruby, Crystal, etc.). However, defining class methods is done with the `static` keyword before a method.

Going with `deftype` and `defstatic` sort of implies heading toward `defmodule` as well, which I would be okay with. Making keywords slightly more obscure is beneficial for end-user code that wants to use names like `module`, `type`, and `static`.

**Comment by faultyserver** [comment]
Because I'm a little stuck and have some time, I'm going to try to implement this most recent solution. I'll also make the change to `defmodule` as part of it.

**Comment by faultyserver** [comment]
One question I came up with last night was how to allow dynamic type specification for initialization. In Ruby, you can do `anything.new`, where `anything` can be any type of value, a block, a variable, a constant, etc.

Here, with the brace-initialization syntax, that luxury doesn't really exist. With the brace-block syntax, something like `type{ }` would be ambiguous.

Arguably, that could be okay, where a type essentially acts as a Callable and the braces just act like a regular block, but I think that would become ugly. With that, you'd be forced to specify variables and stuff for each initialization:

```ruby
deftype Thing
  name : String
  color : String
end

type = Thing

type{
  name  = "barn"
  color = "red"
}
```

A better solution to me would be to allow interpolations as part of the syntax:

```ruby
<type>{ "barn", "red" }
```

To me, that makes it more obvious that a type is expected and being dynamically specified.

**Comment by faultyserver** [comment]
Looking more at the `deftype` syntax, having the properties be defined as part of the body of the type is a little odd. In most `Expressions` (Def bodies, When bodies, etc.), having type definitions doesn't make sense, and I feel like it should be disallowed. Enforcing that would mean having a special Node type for type definitions (and potentially modules as well, if they should be allowed to define properties in this way).

Something I came up with, almost in the same vein as [NamedTuples in Crystal](https://crystal-lang.org/docs/syntax_and_semantics/literals/named_tuple.html) is to move the property definitions for a type into a separate block-like section:

```ruby
deftype Person{name : String, birthday : DateTime, age : Integer}
  defstatic create_twin(p : Person, name : String)
    Person{ name, p.birthday, p.age }
  end
  
  def greeting
    "Hello, " + @name
  end  
end
```

These could also be split onto multiple lines, if that makes the definition more clear:

```ruby
deftype Person{
  name : String,
  birthday : DateTime,
  age : Integer
}
  defstatic create_twin(p : Person, name : String)
    Person{ name, p.birthday, p.age }
  end
  
  def greeting
    "Hello, " + @name
  end  
end
```

I think that becomes a little more odd with regard to visual alignment of the type body (where `defstatic` starts), but it is nicer for types with many properties.

-------

Something else worth mentioning. I think the _properties_ of a type should not necessarily limit the _instance variables_ for the type. The properties would be the public attributes, kind of the equivalent of using `attr_accessor` in Ruby, while other instance variables (likely created on-the-fly with `@<var_name>`) will be private by default (also as in Ruby).

Properties could also be used as the default initialization and pattern matching structure. Instead of a `def initialize`, using the brace-initialization syntax would allow for setting the properties either in order or by name (e.g., `Person{"Jon", D1995-11-14, 21}`). And in pattern matching, only the properties would be matched and/or allowed in the pattern definition (see `same_name?` from above). I don't know how/if a callback for initializing instance variables would be good/useful.

**Comment by faultyserver** [comment]
I can't say I'm completely on board with the property specifications now. While it's nice to explicitly state the structure of the data that the type contains, having it as a distinct part of the declaration isn't very useful, and would really just lead to some confusion about how properties and instance variables might be different.

**Comment by faultyserver** [comment]
Allowing instance variables (or maybe all variables in general) to be typed seems like a beneficial idea, though:

```ruby
deftype Person
  @name : String
  @birthday : DateTime
  @age : Integer
end
```

I think that's worthy of a separate issue/discussion. Being able to assert types for variables at the language level is something I've always felt Ruby lacked, even as a dynamically typed language, especially around enforcing (non-)nilability.

**Comment by faultyserver** [comment]
After messing around in IRB, I realized that one of my assumptions of how Ruby works is wrong. Instance variables defined inside a class but outside of a method are not properties of the instance. For example:

```ruby
class Foo
  @prop = "hello"
  attr_accessor :prop

  class << self
    attr_accessor :prop
  end
end

Foo.new.prop #=> nil
Foo.prop #=> "hello"
```

Having tested this, it now seems pretty obvious that instance variables act on the current value of `self`, which inside a class definition is the class itself.

In any case, that simplifies a lot of what I was thinking needed to be done to support instance variables, and avoids the need for any of the special syntax shown above.

With that, we're basically back at Ruby's implementation of class properties solely being through instance variables defined in methods.

**Comment by faultyserver** [comment]
There's a lot of parsing ambiguity with the `Type{}` initialization syntax. It's unlikely, but consider the case where a user has saved a method into a Const and wants to invoke it with a block argument:

```ruby
SOME_PROC { |x| x + 1 }
```

There's no way for the parser to differentiate this from what would be considered a type initialization such as:

```ruby
Person{ "Jon", 21 }
```

Sadly, they are lexically identical except for the content between the braces, which is not sufficient for clearing the ambiguity when parsing.

To resolve this, for now, I'd like to follow the Elixir style of prefixing struct initialization with a `%`:

```ruby
%Person{ "Jon", 21 }
```

I think this should still allow interpolations for dynamic type instantiation:

```ruby
%<type>{ a, b, c }
```

Much as it's a little awkward and _potentially_ surprising to users, I still like the brace syntax to distinguish creating a new instance from regular method calls. That said, the interpreter should end up calling some constructor method (name TBD) that users can define to properly initialize properties.

**Comment by faultyserver** [comment]
This has been implemented as of #16.
